User only methods

code1mountain opened the pull request:
This PR contains a bunch of user only methods.

Plans for the future (Not this PR, they need many new classes):

- Working with the admin log
- Working with group/channel stats

The new methods will be documented in a machine readable format, probably `openAPI`, so we can generate a Swagger documentation. It doesn't make sense to document them in the readme, which becomes too messy.

### Progress tracking
- [x] ~~`getInlineQueryResults` (sending an inline query to a bot)~~ (A lot of extra code, moved to a future PR)
- [x] `getChats` (Normal chat list)
- [x] `getCommonChats` (Common chats with an other user or bot) * (There is currently a bug in tdlight which causes this method to fail, hopefully it will be fixed soon.)
- [x] `getInactiveChats` (When reaching the limit of available supergroup chats)
- [x] `getNearbyChats` (Search chats by location)
- [x] `searchPublicChats` (Searches public chats by looking for specified query in their username and title.)
- [x] `votePoll` (For voting in polls or quizzes and retracting votes)
- [x] `joinChat` (join a chat by username/chat id or invite_link)
- [x] ~~`joinChatByInviteLink` (join a chat by invite link)~~
- [x] `createChat` (Create a group, supergroup or channel)
- [x] `addChatMember` (add a user to a group or channel)
- [x] `getCallbackQueryAnswer` (Interacting with inline Keyboards)
- [x] ~~`sendInlineQueryResultMessage` (selecting an inline query to send)~~ (A lot of extra code, moved to a future PR)
- [x] `searchMessages` (for searching messages globally)
- [x] `searchChatMessages` (for searching messages in a chat)
- [x] `deleteChatHistory` (delete Chat History in a private chat, group or channel)
- [x] `reportChat` (report a group or a user for spam)
- [x] documentation

If I miss any important methods, please let me know. I'm also open for better name suggestions or other comments.

giuseppeM99 commented:
about the joinchat method
since this is an abstraction layer i think that it would be better to have the same method both for public and private chats

code1mountain commented:
@giuseppeM99 Okay. So the joinChat method takes a chat_id parameter which can be a numerical id or the username (like everywhere in the api). So it is possible to join nearby chats by chat_id as well. I could add an invite_link parameter, that is exclusive to the chat_id, but I'm not sure if this is the nicest solution. I'm also not sure what to do if someone enters something like "t.me/tdlightchat" as invite_link, we should probably fail this query cause there are dozens of alias domains for t.me

andrew-ld commented:
@code1mountain 

the bot api unifies many methods into one, i would make joinchat be able to handle either chat_id (username / id) or joinlink, so it will have two parameters (chat_id and joinlink) and you must enter at least 1 in the request then:

joinchat?chat_id=xxxx
joinchat?joinlink=xxxx

JosXa commented:
Some naming suggestions:

> * [x]  `getChats` (Normal chat list)

`getDialogs`? Pyrogram and Telethon both introduced the concept of a `Dialog` to represent "something in your chats list". I think it makes sense, but up for discussion. It doesn't solve the issue that a channel technically isn't a "chat" though, because "dialog" doesn't make that distinction clearer.

> * [x]  `votePoll` (For voting in polls or quizzes and retracting votes)

There is no real *action* in there. Alternatives:
- `voteInPoll`
- `answerPoll`

On top of that, should we fare differently with quizzes and polls? MTProto considers a quiz a special type of poll and we also can't check the type before making the request (thanks @code1mountain for those infos), but from a user view maybe it'd be useful when they search for "quiz" and find everything relevant in the list.

> * [x]  `joinChat` (join a chat by username or chat id) * (Should we unify these 2 methods to join groups?)
> * [ ]  `joinChatByInviteLink` (join a chat by invite link)

I agree that those should be unified and parametrized.

> * [ ]  `createSupergroupChat` (create a supergroup or channel)

From a user view, I would like to have:
- `createSupergroup`
- `createPrivateGroup`
- `createChannel`

Those do not include the word "chat" however, so you won't find them when searching for that.

> * [ ]  `sendCallbackData` (Interacting with inline Keyboards)

How about `click` or `clickInlineKeyboardButton`?

> * [ ]  `sendInlineQuery` (sending an inline query to a bot)
> * [ ]  `sendInlineQueryResult` (selecting an inline query to send)

Both say "send" but they're kind of different 🤔 

Should one be `requestInlineQueryResults`? Hmm, also sounds weird...

> * [ ]  `searchMessages` (for searching messages in a chat)

Means we will also get `searchChats` (or `searchDialogs`) later, right? Is this for search inside a chat or globaly?

Do we need `searchGlobal` that will combine like the clients do?

giuseppeM99 commented:
@JosXa dialogs are so low level they're not even mentioned in tdlib's methods, as this is an abstraction layer built on top of tdlib, i don't think we should go deeper and re-introduce another concept of dialog that will inevitably be different than the low level mtproto counterpart

JosXa commented:
@giuseppeM99 Ah, I wasn't aware! That's why the other mtproto libs have it, understood.

code1mountain commented:
@andrew-ld The joinchat method only returns ok, while the joinChatByInviteLink returns a chat object. Should I unify it by returning a chat object anyway?

andrew-ld commented:
@code1mountain yes

luckydonald commented:
>> - sendCallbackData (Interacting with inline Keyboards)
> How about click or clickInlineKeyboardButton?

I would go with `sendCallbackData`, as the Bot API also calls it callback data everywhere.
I'd just have the endpoint description say "Clicks a button" instead.

luckydonald commented:
>> -  createSupergroupChat (create a supergroup or channel)
>
> From a user view, I would like to have:
>
> `createSupergroup`
> `createPrivateGroup`
> `createChannel`
> Those do not include the word "chat" however, so you won't find them when searching for that.

how about an unified
#### `createChat`
Field | Type | Description
-- | -- | --
type | String | Type of chat, can be either “private”, “group”, “supergroup” or “channel”

this would simplify the search,  and also be consistent with the [`Chat` object](https://core.telegram.org/bots/api#chat).

JosXa commented:
> > > * sendCallbackData (Interacting with inline Keyboards)
> 
> > How about click or clickInlineKeyboardButton?
> 
> I would go with `sendCallbackData`, as the Bot API also calls it callback data everywhere.
> I'd just have the endpoint description say "Clicks a button" instead.

Thinking about it, as we cannot emulate "clicking on a URL button", I agree with centering this around callback data. However what's weird to me is that "sending" is usually around messages, not machine-specific implementation details. But then since there's no precedence for verbs like `fireCallback`, `executeCallback` or `invokeCallback`, maybe it's better to break consistency instead.

JosXa commented:
> > > * sendCallbackData (Interacting with inline Keyboards)
> 
> > How about click or clickInlineKeyboardButton?
> 
> I would go with `sendCallbackData`, as the Bot API also calls it callback data everywhere.
> I'd just have the endpoint description say "Clicks a button" instead.

Thinking about it, as we cannot emulate "clicking on a URL button", I agree with centering this around callback data. However what's weird to me is that **"sending" is usually around messages**, not machine-specific implementation details. Therefore, I would urge to find a name similar to `fireCallback`, `executeCallback` or `invokeCallback`, as breaking that consistency would be very unintuitive.

It is named "the data of a callback" (aka payload) after all, so "calling back" is what this method should represent.

JosXa commented:
Same should go for inline queries I feel, where "sending" is absolutely valid for an individual inline query result, but firing the `@inline query` should absolutely be named with the same verb that we decide to use for callback_data.

luckydonald commented:
With `answerCallbackQuery` being how the bot acts on them maybe `askCallbackQuery`?

code1mountain commented:
> how about an unified
> 
> #### `createChat`
> Field	Type	Description
> type	String	Type of chat, can be either “private”, “group”, “supergroup” or “channel”
> this would simplify the search, and also be consistent with the [`Chat` object](https://core.telegram.org/bots/api#chat).

Creating a basic group requires a title and a list of members to be added (Not sure if it can be an empty list, but I assume).

Creating a supergroup or channel requires a title, a description (may be empty) and optionally a location. 

So a `createChat` Method would have these parameters:
- `type`: one of `group`, `supergroup`, `channel`
- `title`: String title
- `longitude`/`latitude`: location parameters, only works for supergroups

Adding members to a group can be done later, it's also not the best way to get people to join your group.

code1mountain commented:
For `ReportChat` I would have a string parameter called `reason`. If it is equal to some predefined values, it will default to the Copyright, Spam, Violence... reasons. If it isn't equal to one of these default Values, it will send a (chatReportReasonCustom)[https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1chat_report_reason_custom.html] with the provided string as text.

JosXa commented:
> joinchat?joinlink=xxxx

~joinlink~ -> **invite_link** ?

JosXa commented:
@code1mountain 

> Creating a basic group requires a title and a list of members to be added (Not sure if it can be an empty list, but I assume).

[In the native clients that's not possible](https://s3.nl-ams.scw.cloud/josxa-content/50/FlHCKMiP60.mp4). Can someone confirm using raw mtproto?

giuseppeM99 commented:
as far as i remember, you can't create an empty group
you can however create empty supergroups (megagroups)

luckydonald commented:
So we needs something like a `initial_members` array.
Would that just be the same stuff as the usual `chat_id`s, so either int id or @username?

Should we support the same argument for other groups, too?
While we could only accept it for groups, I bet someone would like to make a fork enabling support for that anyway if we don't.

code1mountain commented:
For `addChatMembers`, it is only possible to add multiple chat members at once in Supergroups/channels. If this method is used in normal groups, it will just add the users one by one. I think that's better than having a separate Method `addChatMember`, which is only callable for basic group chats

luckydonald commented:
`addChatMembers`: Maybe we should limit the given array to exactly one in the case of supergroups/channels, as otherwise errors would be difficult to communicate.

code1mountain commented:
@luckydonald Currently it adds all users recursively in normal groups and returns ok on success. If it fails after a few users, there is currently no way to see where it failed.

luckydonald commented:
@code1mountain I think that's something which should be solve properly.

I'm can think of two ways:

1. Add error handling telling which entry of the list were successfully, which one failed and which are still unprocessed
2. allow only to add users one at a time (for all those similar functions)
   While it would mean having to call the API more often,
    - it should be a lot easier to implement for us
    - fits the overall error structure of having either a success or a fail for one action.
    - it discourages to bulk add users which both can easily get your account limited and generally isn't nice anyway.

The longer I think about it, the more it makes sense that the `members` attribute on that would be only valid for the group, and otherwise users would have to be added one-by-one with a `addChatMember` method.

code1mountain commented:
@luckydonald Okay, makes sense. I think I will change it to addChatMember and only allow a single user at a time to discourage mass adding.

mugavri commented:
If you want I also added some functions:
getChatStats
getStatsGraph
getMessageStats
getMessagePublicForword
getChatEventLog

JosXa commented:
@mugavri you could open a PR onto the branch of this one

code1mountain commented:
@mugavri 
> Plans for the future (Not this PR, they need many new classes):
> 
> * Working with the admin log
> * Working with group/channel stats

I don't want to add these methods in this Pull Request, because it already has a few hundred lines of code and working with stats and recent actions requires some completely new Json classes. Maybe I will add it in a future PR, but you can also add them on your own and send a separate Pull Request.

mugavri commented:
@code1mountain I have already added everything that needed to be added,

code1mountain commented:
How to unify the `deleteChat` method? There are two different things it can do:
1. Delete all Messages in the Chat History (Works for all kinds of chats, but there are differences if the messages are only deleted for yourself or for everybody))
2. Delete a Supergroup chat (removes all members, only works for up to 1000 members, owner only)

so it would have the following parameters:
`chat_id` (obviously)
`only_for_me` (This is inverted from the tdlib implementation, but makes more sense for programming bots. Doesn't work in supergroup chats)
`remove_all_members` (This completely deletes a supergroup chat)

Do you have any better name suggestions?

luckydonald commented:
Maybe "Delete all Messages in the Chat History" shouldn't be the same thing as deleting the chat?
Sound like `deleteChat` and `clearChat` could be a viable separation?
If it can't be deleted (only can be left) it maybe should return an error.

code1mountain commented:
I removed the two methods to work with inline queries from the list, because they would need a lot of extra code and we did not come to a clear conclusion how to name them. Seeing that I already have +575 lines of new code, I will close the list for this pull request. I think that I added the most important methods for userbots.

For the documentation I will write an openAPI file with all the new methods. We can use [Github Pages](https://github.com/peter-evans/swagger-github-pages) to generate and host a nice-looking Swagger documentation from that. It is also machine-readable, so it is possible to automatically generate wrappers for our api. It also makes updating a lot easier.

code1mountain commented:
There's also still the tdlight issue with getCommonChats. Can one of the team members please look into that?

luckydonald commented:
Regarding the swagger docs, Is there a way to have users fill in their own host there dynamically?
Do we need to serve CORS headers for that Ajax stuff?

Or could we maybe even make the API itself serve those at a specified port?
That would link the documentation to the version running, and in case of it going out of date it wouldn't cause so much confusion.

code1mountain commented:
@luckydonald 1. Yes, it is possible to let users enter their own host dynamically

2. No extra headers required afaik

3. It is possible to have multiple yaml files with different documentations for different versions. They could all have their own endpoint on an older version.

luckydonald commented:
@code1mountain if we do 3., we'll need an endpoint serving the version of the api, and reasonable version bumps.

Sounds like just bundling a js file and the openapi directly would be a simpler approach

code1mountain commented:
@luckydonald I did the documentation (for all of our changes, not just from this PR. I'm just not sure about getMemoryStats). You can see it at https://code1mountain.github.io/tdlight-telegram-bot-api-docs/

We can also change the icons of the documentation if we want.
I would still leave a changelog in the readme section, but a lot less detailed.

Please try the new methods and give feedback on the documentation.

MarcoBuster commented:
About the documentation, please discuss it in #14. 
If we decide to adopt a new documentation system like Swagger (which I suggest to) we must deploy it *before* merging this PR.

luckydonald commented:
@code1mountain Maybe searchMessages and searchChatMessages could be unified, with an optional chat_id parameter?

code1mountain commented:
@luckydonald The problem is that when searching globally, some filters are not available. Also it would get quite messy with the from_user_id parameter only working in searchChatMessages and the time restrictions only for searchMessages. I think it is nicer if they are two separated methods

JosXa commented:
> @code1mountain if we do 3., we'll need an endpoint serving the version of the api, and reasonable version bumps.
> 
> Sounds like just bundling a js file and the openapi directly would be a simpler approach

Or just nginx reverse proxy through

luckydonald commented:
@JosXa you just throwing in a `use nginx` seems unrelated to the point I'm making.

JosXa commented:
> @JosXa you just throwing in a `use nginx` seems unrelated to the point I'm making.

Okay, in more detail @luckydonald:

1. In GH actions, configure [a linter](https://github.com/marketplace/actions/openapi-lint-action)
2. Set up a [swagger UI action](https://github.com/marketplace/actions/swagger-ui-action) to create HTML from the OpenAPI docs - optionally with environment specifier (/subpath?)
3. Configure an Nginx container as reverse proxy with SSL termination and path-based routing. There are tons of examples out there :) [Caprover](https://caprover.com) makes it easy, that's probably what I'm gonna use. But I don't know what this "official" host here is working with 🤔

luckydonald commented:
Ah sorry, for serving the `docs/` folder.
Yeah I guess it would work to build the static html somewhere.

But accessing those file from a different (nginx) container would be quite cumbersome (is it even possible?).
Having two containers for that and having to update two containers all the time feels error prone, too.

Maybe whipping up a port similar to the stats one (maybe a `--http-docs-port=4458` or something) would solve that in my eyes. This could then easily proxied with any reverse proxy or even accessed without.

luckydonald commented:
I think it is quite a great to have the documentation served by the same hosts the api runs on, as that gives the freedom for forks to add custom stuff easily or work with that one version you deemed working fine and don't wanna update all the time.

With the api e.g. running at https://telegram.rest, simply going to https://telegram.rest/docs makes it easy for the end user.

JosXa commented:
It's a very, *very* standard thing to share the same network between two (three) docker-compose services (or the equivalent in docker swarm) with one running a raw nginx image and the other serving your API, and then depending on how you like it a third one serving the docs or nginx just serving static files.

JosXa commented:
@luckydonald Yes the port idea is great. But someome will actually have to make a server, right? As I understood the user files also just sit on disk and are supposed by sth like jenkins - so no change in that regard actually.

code1mountain commented:
@JosXa 
> About the documentation, please discuss it in #14. 
> If we decide to adopt a new documentation system like Swagger (which I suggest to) we must deploy it *before* merging this PR.

code1mountain commented:
I added an alias `authpassword` for `2fapassword`, because `2fapassword` is not a valid name for functions or methods in many programming languages and I want to simplify the naming of new functions. Only `authpassword` is mentioned in the docs, but `2fapassword` still works.

code1mountain commented:
I will not add any new features to this pull request and the documentation is pretty much ready (just some discussion on how/where to deploy it, that's why I have it [here](https://code1mountain.github.io/tdlight-telegram-bot-api-docs/) for now).

Please review and test the code, so we can deploy the documentation and this Pull Request at the same time?

PS: Can someone please help me merging the conflicts? Never done this before.

luckydonald commented:
@code1mountain Please make sure it's authPath and the same camelCase for the login in the docs, so we stay consistent

MarcoBuster commented:
(Let's not use [WIP] in titles, use the "draft PR" feature)

MarcoBuster commented on telegram-bot-api/Client.cpp:
```diff
@@ -4043,7 +4261,7 @@ void Client::on_update_authorization_state() {
       if (is_user_) {
         return send_request(make_object<td_api::setAuthenticationPhoneNumber>(phone_number_, nullptr),
-            std::make_unique<TdOnAuthorizationCallback>(this));
+                            std::make_unique<TdOnAuthorizationCallback>(this));
```
Useless indentation

code1mountain replied:
Outdated now

code1mountain commented:
I added message scheduling to this commit, because it took longer to merge than expected and I used some code from the other added methods. All `send` methods have now an optional parameter `send_at`, which can either be a unix timestamp or `online` (to send when the other chat participant comes online).

Scheduled messages have a negative `message_id` to destinct them from normal message_ids. `Message` has 2 new optional parameters now, `is_scheduled` and `scheduled_at`. There are also 2 new methods `getScheduledMessages` and `editMessageScheduling`.

The existing methods `editMessageText`, `editMessageCaption`, `editMessageMedia` and `deleteMessage` also work exactly the same to edit/delete scheduled messages. I will add the documentation to my other PR soon.

giuseppeM99 commented on telegram-bot-api/Client.cpp:
```diff
@@ -6008,9 +6289,9 @@ td::int64 Client::get_message_id(const Query *query, Slice field_name) {
-  if (arg < 0) {
-    return 0;
-  }
+  // if (arg < 0) {
```
Why did you remove this check?

code1mountain replied:
Scheduled messages have negative message ids. Therefore this check if the message_id is negative breaks scheduled messages.

giuseppeM99 replied:
Ok, but this could possibly break other methods that expect the message_id to be positive, especially with bots that can't schedule messages, or methods that should work with non scheduled messages, adding a parameter to allow negative numbers would be fine for me

code1mountain replied:
The check basically moved to as_tdlib_message_id. There it uses the old conversion (20 bit shift) only for positive message_ids. It still has the same check for positive numbers and an other behaviour for negative ones.

andrew-ld replied:
arg < 0 && is_bot

code1mountain replied:
Okay, I will add that.

giuseppeM99 commented on telegram-bot-api/Client.cpp:
```diff
@@ -7704,6 +8050,232 @@ td::Status Client::process_get_memory_stats_query(PromisedQueryPtr &query) {
+    send_request(make_object<td_api::createNewBasicGroupChat>(std::move(initial_members), title.str()),
+                 std::make_unique<TdOnReturnChatCallback>(this, std::move(query)));
+  }
+  return Status::OK();
```
Handle the case when chat_type is something else

code1mountain replied:
It raises an error now, done

giuseppeM99 commented on telegram-bot-api/Client.cpp:
```diff
@@ -7704,6 +8050,232 @@ td::Status Client::process_get_memory_stats_query(PromisedQueryPtr &query) {
+  auto title = query->arg("title");
+
+  if (chat_type == "supergroup") {
+    send_request(make_object<td_api::createNewSupergroupChat>(title.str(), false, "", nullptr),
```
Add an optional description parameter

code1mountain replied:
I decided not to have that parameter, because it can only be set for supergroups or channels. Basic Groups can't have a description. I wanted to keep them as similar as possible and a description can still be set in a separate method call. Having 2 ways to set a description for a group is bad api design IMO

giuseppeM99 replied:
So one would have to make 2 api calls to create a channel with a description? I'd just have it as an optional parameter, we could just ignore it for normal group chats and assume empty if not given

code1mountain replied:
Okay, I will add that.

giuseppeM99 reviewed (changes requested):
There are also some formatting issues, use clang-format to fix them automatically

code1mountain commented on telegram-bot-api/Client.h:
```diff
@@ -101,6 +101,8 @@ class Client : public WebhookActor::Callback {
   class JsonChatPhotoInfo;
   class JsonChatLocation;
   class JsonChat;
+  class JsonChats;
```
Okay, good idea. I will do that.

code1mountain replied:
There are also a bunch of custom helper methods mixed throughout the code, like the disable_internet_connection, get_int_array_arg... Can I also move them together after the upstreams, even if they are out of the scope of this (already massive) pull request?

giuseppeM99 replied:
Yes move them

code1mountain replied:
Everything is moved

code1mountain commented on telegram-bot-api/Client.cpp:
```diff
@@ -3349,6 +3456,129 @@ class Client::TdOnGetMemoryStatisticsCallback : public TdQueryCallback {
+  PromisedQueryPtr query_;
+};
+
+class Client::TdOnJoinChatIdCallback : public TdQueryCallback {
```
Okay, makes sense. I first thought I would need an extra callback for joining by invite link, but it just returns a chat, so I could reuse an other callback. Can clean that up now.

code1mountain replied:
It's renamed now, done.

code1mountain commented on telegram-bot-api/Client.cpp:
```diff
@@ -788,6 +851,24 @@ class Client::JsonMessages : public Jsonable {
   const td::vector<td::string> &messages_;
 };
 
+class Client::JsonMessages_ : public Jsonable {
```
I renamed it to JsonMessagesArray

MarcoBuster commented:
@code1mountain what's the status of this? Are you just waiting for our review?

code1mountain commented:
Yes, mainly waiting for review.
